include/deal.II/lac/solver_gmres.h

```diff
@@ -710,81 +723,319 @@ SolverGMRES<VectorType>::givens_rotation(Vector<double> &h,
 
 
 
-template <class VectorType>
-inline double
-SolverGMRES<VectorType>::iterated_modified_gram_schmidt(
-  const internal::SolverGMRESImplementation::TmpVectors<VectorType>
-    &                                       orthogonal_vectors,
-  const unsigned int                        dim,
-  const unsigned int                        accumulated_iterations,
-  VectorType &                              vv,
-  Vector<double> &                          h,
-  bool &                                    reorthogonalize,
-  const boost::signals2::signal<void(int)> &reorthogonalize_signal)
+namespace internal
 {
-  Assert(dim > 0, ExcInternalError());
-  const unsigned int inner_iteration = dim - 1;
+  namespace SolverGMRESImplementation
+  {
+    template <class VectorType>
+    void
+    dot_and_add(
+      const unsigned int dim,
+      const VectorType & vv,
+      const internal::SolverGMRESImplementation::TmpVectors<VectorType>
+        &             orthogonal_vectors,
+      Vector<double> &h)
+    {
+      for (unsigned int i = 0; i < dim; ++i)
+        h[i] += vv * orthogonal_vectors[i];
+    }
 
-  // need initial norm for detection of re-orthogonalization, see below
-  double     norm_vv_start = 0;
-  const bool consider_reorthogonalize =
-    (reorthogonalize == false) && (inner_iteration % 5 == 4);
-  if (consider_reorthogonalize)
-    norm_vv_start = vv.l2_norm();
 
-  for (unsigned int i = 0; i < dim; ++i)
-    h[i] = 0;
 
-  for (unsigned int c = 0; c < 2; ++c) // 0: orthogonalize, 1: reorthogonalize
+    template <class Number>
+    void
+    dot_and_add(
+      const unsigned int                                                   dim,
+      const LinearAlgebra::distributed::Vector<Number, MemorySpace::Host> &vv,
+      const internal::SolverGMRESImplementation::TmpVectors<
+        LinearAlgebra::distributed::Vector<Number, MemorySpace::Host>>
+        &             orthogonal_vectors,
+      Vector<double> &h)
     {
-      // Orthogonalization
-      double htmp = vv * orthogonal_vectors[0];
-      h(0) += htmp;
+      AlignedVector<VectorizedArray<double>> hs(dim,
+                                                VectorizedArray<double>(0.0));
+
+      unsigned int j = 0;
+      for (unsigned int c = 0;
+           j < vv.locally_owned_size() / VectorizedArray<double>::size();
+           ++c, j += VectorizedArray<double>::size())
+        for (unsigned int i = 0; i < dim; ++i)
+          DEAL_II_OPENMP_SIMD_PRAGMA
+      for (unsigned int v = 0; v < VectorizedArray<double>::size(); ++v)
+        hs[i][v] +=
+          orthogonal_vectors[i].local_element(j + v) * vv.local_element(j + v);
```

Comment:
Is this what you say in the other comment: You want to do the accumulation in `double` variables? I think this should work, but I think it would be better to cast to `double` already for the product.

Reply:
Another comment on this code: Here, we load from `dim` different vectors in a linear fashion. At some point in the past, there were cases when involving more than x different streams (with x between 10 and 20) would lead to misses in the translation lookaside buffer (when the CPU translates virtual memory addresses as seen by the pointers to physical addresses in memory). I think nowadays CPUs are capable enough that one would never see this, but I haven't checked the TLB capabilities for a while.

Reply:
> I think this should work, but I think it would be better to cast to double already for the product.

I am not sure about this. I guess we also support complex numbers (maybe not for the distributed vector case but in the normal vector case). That would not work or is there a way to cast that?

Reply:
Our solvers do not support complex numbers as far as I can tell from the stored matrices and vectors of the class. I recommend reading section 6.5.9 "GMRES for complex systems" in the book by Saad (Iterative methods for sparse linear systems, 2nd edition). Several of those should be complex-valued in case we actually supported the complex case. I am not saying we shouldn't do, but that would be more work and should be done by someone having interest in those settings.